### PR TITLE
Symmetrization of Wannier Hamiltonian and matrices

### PR DESCRIPTION
**Pull request by Liu-Xiaoxiong**
1. Symmetrisation of Ham and Vector matrix (eg. AA BB SS ...), in system_w90 and system_tb. 
   Only Ham including Time-reversal now.
   Working for orbitals s,p,d,f, sp3. (more orbitals in the future)
2. Adding auto sphere k-list in __path.py.

### CODE REVIEW

**Comment by stepan-tsirkin** [comment]
Great @Liu-Xiaoxiong,  could you also add a test? 

Maybe to integrate something using the symmetrized model, with and without using the symmetry during integration. Results should be the same, right? then, no additional reference files are needed.

**Comment by jaemolihm** [comment]
Just out of curiosity (I haven't read the source code), how do you detect the orbital character of each Wannier functions? Do you read the seedname.win file?

Also, for the test, making a model with small symmetry breaking and testing that the model becomes fully symmetric (e.g. bands are degenerate) may be useful, as this will be the typical use case.

**Comment by Liu-Xiaoxiong** [comment]
> Just out of curiosity (I haven't read the source code), how do you detect the orbital character of each Wannier function? Do you read the seedname.win file?

It was reading the seedname.win file, but now we need to manually input the projections which are the same as the projection card in .win file.  Following the same way in which wannier90 arranges orbital characters,  we can mark them for each Wannier function. 

> 
> Also, for the test, making a model with small symmetry breaking and testing that the model becomes fully symmetric (e.g. bands are degenerate, integration results are symmetric) may be useful, as this will be the typical use case.

Thanks, I am going to do that.

**Comment by Liu-Xiaoxiong** [comment]
> Great @Liu-Xiaoxiong, could you also add a test?
> 
> Maybe to integrate something using the symmetrized model, with and without using the symmetry during integration. Results should be the same, right? then, no additional reference files are needed.

Okay, thanks!

File: wannierberri/__sym_wann.py [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+            syr=np.array([[0.0,1.0],[-1.0,0.0]])
+            self.ul=np.kron(syl,base_m)
+            self.ur=np.kron(syr,base_m)
+        
+        #========================================================
+        #symmetrize exist R vectors and find additional R vectors 
+        #========================================================
+        print('##########################')
+        print('Existing Block')
+        res_dic_1, iRvec_add =  self.average_H(self.iRvec,keep_New_R=True)
```
Comment (stepan-tsirkin) [inline wannierberri/__sym_wann.py:515]: one week point about that. In theory, after you find new vectors, if you repeat the procedure, you may find even more Rvectors (the new Rs from previous iteration transformed), until the procedure converges. 

What I would do, is to find All new Rvectors iteratively, then extend all amtrices to the new size (by zeros), and then perform the symmetrization.

**Review by stepan-tsirkin — CHANGES_REQUESTED** [review]
I created an example in ` tests/data/Fe_Wannier90_sym` . When only internal terms (derived from Ham) are needed, the result is nicely symmetric. When it goes to matrices AA, SS etc, it becomes a mess. the spin seems to vanish. @Liu-Xiaoxiong, can you look what is happening?

File: wannierberri/__sym_wann.py [hunk 2]
```diff
@@ -0,0 +1,554 @@
...
+        '''
+        rot_map: A map to show which atom is the equivalent atom after rotation operation.
+        vec_shift_map: Change of R vector after rotation operation.
+        '''
+        wann_atom_positions = [self.wann_atom_info[i][2] for i in range(self.num_wann_atom)]
+        rot_map=[]
+        vec_shift_map=[]
+        for atomran in range(self.num_wann_atom):
+            atom_position=np.array(wann_atom_positions[atomran])
+            new_atom =np.round( np.dot(self.symmetry['rotations'][sym],atom_position) + self.symmetry['translations'][sym],decimals=5)
```
[the remark this answers is not in the document]
Comment (Liu-Xiaoxiong) [reply wannierberri/__sym_wann.py]: No.

File: wannierberri/__sym_wann.py [hunk 3]
```diff
@@ -0,0 +1,554 @@
...
+            for atom_index in range(self.num_wann_atom):
+                old_atom= np.round(np.array(wann_atom_positions[atom_index]),decimals=5)
+                diff = np.array(np.round(new_atom-old_atom,decimals=8))
+                if abs(diff[0]%1)+abs(diff[1]%1)+abs(diff[2]%1)<10E-5:
```
[the remark this answers is not in the document]
Comment (Liu-Xiaoxiong) [reply wannierberri/__sym_wann.py]: Thanks ,done

File: wannierberri/__sym_wann.py (hunk 1 above)
Comment (Liu-Xiaoxiong) [reply wannierberri/__sym_wann.py:614]: Already canceled the Existing Block.

**Comment by Liu-Xiaoxiong** [comment]
Now the sym_wann method works for AA_R Ham_R and SS_R.
I add a test about AHC of GaAs, If the model with respect to the symmetry, AHC should be 0. 
GaAs with [2,2,2] ab-initio grid gives us some peak up to -0.2.    After symmetrization, the highest peak reduces to 1e-9.

Improved the plot_path_fat function.
In the master version. When you plot the berry_curvature. Some points close to the band crossing may be larger than the whole figure. 
  
Now: 
```
if quantity > 2:
      quantity = np.log2(quantity)
else:
      quantity *= 0.5
```
It can keep the continuity and solve the problem.
What do you think??

File: wannierberri/__path.py [hunk 4]
```diff
@@ -31,22 +31,19 @@ class Path(Grid):
 
     """
 
-    def __init__(self,system,k_list=None,k_nodes=None,length=None,dk=None,nk=None,labels=None,breaks=[]):
+    def __init__(self,system,k_list=None,k_nodes=None,length=None,dk=None,nk=None,labels=None,breaks=[],r1=None,r2=None,ntheta=None,nphi=None,origin=None):
```
[the remark this answers is not in the document]
Comment (Liu-Xiaoxiong) [reply wannierberri/__path.py:44]: done

File: wannierberri/fermiocean.py [hunk 5]
```diff
@@ -197,6 +197,8 @@ def spin_hall_ryoo(data_K,Efermi,tetra=False,degen_thresh=1e-4,degen_Kramers=Fal
     return spin_hall(data_K,Efermi,"ryoo",tetra=tetra,**kwargs_formula)
 
 
+
```
[the remark this answers is not in the document]
Comment (Liu-Xiaoxiong) [reply wannierberri/fermiocean.py:199]: Sorry

**Comment by Liu-Xiaoxiong** [comment]
The red cross at "Lint/flake" is my error or in the master? I saw master also have it.

File: wannierberri/__sym_wann.py [hunk 6]
```diff
@@ -0,0 +1,554 @@
...
+        print("[get_spacegroup]")
+        print("  Spacegroup is %s." %spglib.get_spacegroup(cell))
+        self.symmetry = spglib.get_symmetry_dataset(cell)
+        self.nsymm = self.symmetry['rotations'].shape[0]
+        show_symmetry(self.symmetry)
+
+    def get_angle(self,sina,cosa):
+        '''Get angle in radian from sin and cos.'''
+        if abs(cosa) > 1.0:
+            cosa = np.round(cosa,decimals=1)
```
[the remark this answers is not in the document]
Comment (Liu-Xiaoxiong) [reply wannierberri/__sym_wann.py:185]: okay, round can solve the problem when it <-1 and >1.

**Comment by jaemolihm** [comment]
> The red cross in front of "Lint/flake" from alto check is my error or in the master? I saw master also have it.

No, it's not an error on your side.

**Comment by jaemolihm** [comment]
> Now the sym_wann method works for AA_R Ham_R and SS_R. I add a test about AHC of GaAs, If the model with respect to the symmetry, AHC should be 0. GaAs with [2,2,2] ab-initio grid gives us some peak up to -0.2. After symmetrization, the highest peak reduces to 1e-9.

I think the test is not included in this PR. You should update or add the `tests/test_*.py` file that you made. Also, it would be better to use `compare_zero` (in `test_integrate.py`) to test AHC after symmetrization, instead of creating a reference data file.

File: tests/test_integrate.py [hunk 7]
```diff
@@ -256,6 +256,11 @@ def test_GaAs_tb(check_integrate,system_GaAs_tb, compare_energyresult,quantities
                parameters_K = {'_FF_antisym':True,'_CCab_antisym':True},
                   extra_precision = {"berry_dipole_fsurf":1e-6}  )   # This is a low precision for the nonabelian thing, not sure if it does not indicate a problem, or is a gauge-dependent thing
 
+def test_GaAs_sym_tb(check_integrate,system_GaAs_sym_tb, compare_energyresult,['ahc'],Efermi_GaAs):
+    """Test ahc with sym_wann method"""
+    check_integrate(system_GaAs_tb , quantities_GaAs , fout_name="ahc_GaAs_sym_tb" , suffix="" , Efermi=Efermi_GaAs , comparer=compare_energyresult,compare_zero=True)
```
Comment (jaemolihm) [inline tests/test_integrate.py:261]:
```suggestion
def test_GaAs_sym_tb(check_integrate,system_GaAs_sym_tb, compare_energyresult, Efermi_GaAs):
    """Test ahc with sym_wann method"""
    check_integrate(system_GaAs_tb , ["ahc"], fout_name="ahc_GaAs_sym_tb" , suffix="" , Efermi=Efermi_GaAs , comparer=compare_energyresult,compare_zero=True)
```

Comment (jaemolihm) [reply tests/test_integrate.py:261]: The definition of `system_GaAs_sym_tb` in `create_system.py` file is not uploaded.

File: wannierberri/__path.py [hunk 8]
```diff
@@ -20,33 +20,42 @@ class Path(Grid):
         | No labels or nk's should be assigned to None nodes
     nk : int  or list or numpy.array(3) 
         number of k-points along each directions 
-    k_list : array-like
-        coordinatres of all k-points in the reduced coordinates
+    k_list : list or str
+        |if k_list is a list  - Coordinatres of all k-points in the reduced coordinates
+        |if k_list = 'sphere' - Automatically generate k-points on a sqhere (request r1 r2 origin ntheta nphi)
```
Comment (jaemolihm) [inline wannierberri/__path.py:25]:
```suggestion
        |if k_list = 'sphere' - Automatically generate k-points on a sphere (request r1 r2 origin ntheta nphi)
```

File: wannierberri/__path.py [hunk 9]
```diff
@@ -88,7 +97,37 @@ def __init__(self,system,k_list=None,k_nodes=None,length=None,dk=None,nk=None,la
...
+            assert  self.K_list.shape[0]>0, "k_list should not be empty"
+            for var in 'k_nodes','length','nk','dk':
+                if locals()[var] is not None:
+                    warning("k_list was entered manually, ignoring {}".format(var))
+            self.labels={} if labels is None else labels
+            self.breaks=[] if breaks is None else breaks
+        self.div = np.shape(self.K_list)[0]
         self.breaks=np.array(self.breaks,dtype=int)
+  
+    def sphere(self,type_sphere,r1,r2,ntheta,nphi,origin):
```
Comment (jaemolihm) [inline wannierberri/__path.py:112]: What about removing `type_sphere` argument and using `self.sphere(r1,r1,ntheta,nphi,origin)` for the sphere case?

**Comment by jaemolihm** [comment]
The sphere and spheriod mode is interesting. What is your use case? Do you use some custom plotting script to visualize the sphere? If so, it would be useful to add it to the example or even a function (like plot_band_fat).

File: wannierberri/__path.py (hunk 8 above)
Comment (jaemolihm) [reply wannierberri/__path.py:25]: Also, one does not need r2 for the sphere mode.

File: wannierberri/__sym_wann.py [hunk 10]
```diff
@@ -0,0 +1,625 @@
+import numpy as np
+import spglib
+import numpy.linalg as la
+import sympy as sym
+
+#np.set_printoptions(threshold=np.inf,linewidth=500)
+np.set_printoptions(suppress=True,precision=4,threshold=np.inf,linewidth=500)
+
+
+class sym_wann():
```
Comment (jaemolihm) [inline wannierberri/__sym_wann.py:10]: Class names should normally use the CapWords convention. So `SymWann` would be better.
https://peps.python.org/pep-0008/#class-names

File: wannierberri/__sym_wann.py [hunk 11]
```diff
@@ -0,0 +1,625 @@
...
+    '''
+    Symmetrize wannier matrixes in real space: Ham_R, AA_R, BB_R, SS_R,...
```
Comment (jaemolihm) [inline wannierberri/__sym_wann.py:12]:
```suggestion
    Symmetrize wannier matrices in real space: Ham_R, AA_R, BB_R, SS_R,...  
```

File: wannierberri/__path.py (hunk 9 above)
Comment (Liu-Xiaoxiong) [reply wannierberri/__path.py]: Done

**Comment by Liu-Xiaoxiong** [comment]
> The sphere and spheriod mode is interesting. What is your use case? Do you use some custom plotting script to visualize the sphere? If so, it would be nice to add it to the example or even a function (like plot_band_fat).

Sphere subspace is used to show spin texture or berry curvature around  Weyl point or calculate Wilson loop with geometry.  A spheroid is better than a sphere, sometimes, it can avoid some unwanted band crossing in the geometry. 

It can help to check or calculate some results like:
FIG 3 in :[https://journals.aps.org/prl/pdf/10.1103/PhysRevLett.125.216402]
FIG 3 in :[https://arxiv.org/pdf/2104.11254.pdf]

Yes, I am planning to add a plotting function.

File: wannierberri/__sym_wann.py (hunk 10 above)
Comment (Liu-Xiaoxiong) [reply wannierberri/__sym_wann.py]: Done

File: tests/test_integrate.py [hunk 12]
```diff
@@ -153,6 +153,10 @@ def quantities_Haldane():
 def quantities_GaAs():
     return  ["berry_dipole","berry_dipole_test","berry_dipole_fsurf"]
 
+@pytest.fixture(scope="module")
+def quantities_sym_GaAs():
+    return  ["ahc"]
+
```
Comment (jaemolihm) [inline tests/test_integrate.py:159]: Delete this block. No need to use fixture.

```suggestion
```

File: tests/test_integrate.py [hunk 13]
```diff
@@ -276,6 +280,10 @@ def test_GaAs_tb_wcc_ws(check_integrate,system_GaAs_tb_wcc_ws, compare_energyres
     check_integrate(system_GaAs_tb_wcc_ws , quantities_GaAs_internal, fout_name="berry_GaAs_W90" , suffix="tb_wcc_ws",suffix_ref="" , Efermi=Efermi_GaAs , comparer=compare_energyresult ,
                   extra_precision = {"conductivity_ohmic":-2e-6} )   # This is a low precision for the nonabelian thing, not sure if it does not indicate a problem, or is a gauge-dependent thing
 
+def test_GaAs_sym_tb(check_integrate,system_GaAs_sym_tb, compare_energyresult,quantities_sym_GaAs,Efermi_GaAs):
+    """Test ahc with sym_wann method"""
+    check_integrate(system_GaAs_sym_tb , quantities_sym_GaAs, fout_name="ahc_GaAs_tb" , suffix="sym" , Efermi=Efermi_GaAs , comparer=compare_energyresult,compare_zero=True)
+
```
Comment (jaemolihm) [inline tests/test_integrate.py:286]:
```suggestion
def test_GaAs_sym_tb(check_integrate,system_GaAs_sym_tb, compare_energyresult, Efermi_GaAs):
    """Test ahc with sym_wann method"""
    check_integrate(system_GaAs_sym_tb , ["ahc"], fout_name="ahc_GaAs_tb" , suffix="sym" , Efermi=Efermi_GaAs , comparer=compare_energyresult,compare_zero=True)
```

File: wannierberri/__system.py [hunk 14]
```diff
@@ -118,6 +118,33 @@ def set_parameters(self,**parameters):
...
+            except KeyError:
+                pass
+        symmetrize_wann = SymWann(num_wann=self.num_wann,lattice=self.real_lattice,positions=positions,atom_name=atom_name,
+            proj=proj,iRvec=self.iRvec,XX_R=XX_R,soc=soc,magmom=magmom)
+        XX_R,self.iRvec = symmetrize_wann.symmetrize()
+        for X in ['Ham','AA','BB','CC','SS','FF','SA','SHA','SR','SH','SHR']:
+            try:
+                vars(self)[X+'_R'] = XX_R[X]
+            except KeyError:
+                pass
```
Comment (jaemolihm) [inline wannierberri/__system.py:145]:
```suggestion
        for X in keys(XX_R):
            vars(self)[X+'_R'] = XX_R[X]
```

File: wannierberri/__tabulate.py [hunk 15]
```diff
@@ -358,6 +358,9 @@ def plot_path_fat(self,
                     selE=(e<=Emax)*(e>=Emin)
                     klineselE=kline[selE]
                     y=data[selE][:,ib]
+                    select = np.abs(y) > 2
+                    y[select] = np.log2(y[select])
+                    y[~select] *= 0.5
```
Comment (jaemolihm) [inline wannierberri/__tabulate.py:363]: Please add a comment (in the docstring for `plot_path_fat`) that the circle size changes  linearly below 2 and logarithmically above 2.

File: wannierberri/fermiocean.py [hunk 16]
```diff
@@ -196,6 +196,8 @@ def spin_hall_ryoo(data_K,Efermi,tetra=False,degen_thresh=1e-4,degen_Kramers=Fal
...
+
```
Comment (jaemolihm) [inline wannierberri/fermiocean.py:200]: Please undo these changes.

**Review by jaemolihm — COMMENTED** [review]
I made some comments. I do not have the time to review the `__sym_wann.py` file.

**Comment by jaemolihm** [comment]
I suggest to use **squash merging** for this PR because the history is very long. Since most changes are localized to symmetrization routines, there will be little conflicts (hopefully).

File: tests/test_integrate.py (hunk 12 above)
Comment (Liu-Xiaoxiong) [reply tests/test_integrate.py]: Done

File: tests/test_integrate.py (hunk 13 above)
Comment (Liu-Xiaoxiong) [reply tests/test_integrate.py]: Done

File: wannierberri/__system.py (hunk 14 above)
Comment (Liu-Xiaoxiong) [reply wannierberri/__system.py]: Thanks. Done

File: wannierberri/__tabulate.py (hunk 15 above)
Comment (Liu-Xiaoxiong) [reply wannierberri/__tabulate.py:366]: Done

File: wannierberri/fermiocean.py (hunk 16 above)
Comment (Liu-Xiaoxiong) [reply wannierberri/fermiocean.py]: Sorry

**Comment by jaemolihm** [comment]
I merged master and resolved the conflicts due to the tests reorganization.

File: wannierberri/__sym_wann.py [hunk 17]
```diff
@@ -0,0 +1,437 @@
...
+                                        #X_L: only rotation wannier centres from L to L' before rotating orbitals.
+                                        XX_L = self.matrix_list[X][self.H_select[rot_map[atom_a],rot_map[atom_b]],new_Rvec_index,:].reshape(num_w_a,num_w_a,3)
+                                        #special even with R == [0,0,0] diagonal terms.
+                                        if iR == self.iRvec.index([0,0,0]) and atom_a == atom_b:
+                                            if X == 'AA':
+                                                XX_L += np.einsum( 'mn,p->mnp',np.eye(num_w_a),(vec_shift[atom_a]-self.symmetry['translations'][rot]).dot(self.lattice))
+                                        #X_all: rotating vector.
+                                        matrix_list_all[X][iR,atom_a,atom_b,self.H_select[atom_a,atom_b],:] = np.einsum('ij,nmi->nmj',self.rot_c[rot],XX_L).reshape(-1,3)
+                                    else: 
+                                        print (f"WARNING: I do not know how to symmetrize  {X}")
```
Comment (stepan-tsirkin) [inline wannierberri/__sym_wann.py:350]: Or what should the code do for other matrices?

File: wannierberri/__sym_wann.py [hunk 18]
```diff
@@ -0,0 +1,437 @@
...
+                            else:
+                                norm = np.linalg.norm(self.Ham_R[self.H_select[rot_map[atom_a],rot_map[atom_b]],iR])
+                                if norm > 1e-8:
+                                    print (f"WARNING : after symmetry transformation got a new Rvec: {new_Rvec} for iR = {iR}, Rvec = {self.iRvec[iR]} and atoms {atom_a} and {atom_b}"+
```
Comment (stepan-tsirkin) [inline wannierberri/__sym_wann.py:354]: This warning does not happen if `use_ews = True` is used. We could force use_ws before symmetrization, but `System_tb` does not have `mp_grid` if it is not provided manually

**Review by stepan-tsirkin — CHANGES_REQUESTED** [review]
I reorganized the code a bit, separated part into `__symm_wann_orbitals.py`. Could you add a test for a magnetic system (bcc Fe)

**Review by stepan-tsirkin — APPROVED** [review]
Now fantastic!
